File: src/Controllers/CycleController.cs
using Microsoft.Xna.Framework;
using Monocle;
using Celeste.Mod.Entities;

namespace Celeste.Mod.RainTools {
    [Tracked(true)]
    [GlobalEntity]
    [CustomEntity("RainTools/CycleController")]
    public class CycleController : Entity {
        public string CycleTag;
        public float Offset, SecondsPerCycle;
        public string Flag;

        public CycleController(EntityData data, Vector2 offset) : base() {
            Tag |= Tags.Global;

            CycleTag = data.Attr("cycleTag");
            Offset = data.Float("progressionOffset");
            SecondsPerCycle = data.Float("secondsPerCycle");

            if (data.Bool("transitionUpdate"))
                Tag |= Tags.TransitionUpdate;
            if (data.Bool("frozenUpdate"))
                Tag |= Tags.FrozenUpdate;
            if (data.Bool("pauseUpdate"))
                Tag |= Tags.PauseUpdate;

            if (Offset != 0)
                Depth = -2;
            else
                Depth = -1;

            Flag = data.Attr("flag");
        }

        public override void Update() {
            base.Update();

            if (Flag != "" && !(Scene as Level).Session.GetFlag(Flag))
                return;

            if (!Cycles.Has(CycleTag))
                Cycles.SetProgression(CycleTag, Offset);

            if (SecondsPerCycle > 0)
                Cycles.AddProgression(CycleTag, Engine.DeltaTime / SecondsPerCycle);
        }
    }
}

File: Loenn/entities/cycle_controller.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")
local drawableLine = require("structs.drawable_line")

local cycle_controller = {}

cycle_controller.name = "RainTools/CycleController"
cycle_controller.texture = "RainTools/cycle_controller"
cycle_controller.depth = 0
cycle_controller.placements = {
  name = "controller",
  data = {
    cycleTag = "",
    progressionOffset = 0,
    secondsPerCycle = 0,
    transitionUpdate = true,
    frozenUpdate = true,
    -- pauseUpdate = false, -- todo add this later once it doesn't look bad
    flag = "",
  }
}
cycle_controller.fieldOrder = {
  "x", "y",
  "cycleTag",
  "progressionOffset", "secondsPerCycle",
  "transitionUpdate", "frozenUpdate", -- "pauseUpdate",
  "flag"
}

return cycle_controller
